_posts/2023-07-xx-scala-as-a-junior-dev.md

```diff
@@ -0,0 +1,510 @@
+---
+title: "Scala as a junior developer"
+date: 2023-07-xx
+header:
+  image: ""
+tags: []
+excerpt: "My beginner's story with Scala"
+toc: true
+toc_label: "Table of contents"
+---
+
+## 1. Why Scala?
+
+_"Scala is too abstract", "Scala is not for beginners", "Functional programming is like magic", etc._
+
+These are common arguments that come up when discussing Scala or functional programming in general. While some of them may be partially true, it is precisely these reasons that make Scala so interesting to me. Because of its abstraction and expressiveness capabilities, the safety brought by functional programming, and the modularity brought by object-oriented programming.
+
+### 1.1. Benefits of functional programming
+
+I see several benefits in using a functional programming style. Here are a few of them:
+
+#### 1.1.1. Predictability
+
+In functional programming, the result of a method only depends on the arguments passed to it, regardless of the state of the program. This helps to decouple the code, making it easier to test, maintain, and debug. For instance, let's say we want to add 1 to all the elements of a list:
+
+```scala
+// result depends on the argument
+var list = List(1, 2, 3)
+def add1ToAll(l: List[Int]): List[Int] = l.map(x => x + 1)
+add1ToAll(list) // List(2, 3, 4)
+list = List(2, 3, 4)
+add1ToAll(list) // List(3, 4, 5)
+```
+
+With this functional programming compliant implementation, I can predict the result without even computing it, because it depends only on the list I feed it with. I could even replace the function call with its result, and it would still work as expected.
+
+However, if I do not comply with functional programming principles, I can't be sure of the result without computing it, because it does not depend on the input anymore, but on the state of the program, which can change at any time:
+
+```scala
+// result depends on the state of the program
+var list = List(1, 2, 3)
+def add1ToAll(): List[Int] = list.map(x => x + 1)
+add1ToAll() // List(2, 3, 4)
+// list is changed somewhere else in the program
+add1ToAll() // List(3, 4, 5)
+```
+
+This produces non-deterministic code with side-effects, which can lead to bugs that are hard to find because they might appear only after a very specific and/or long sequence of events.
```

Comment:
The program is not exactly non-deterministic, but I get what you wanted to say.
I think a better framing would be: because we have side effects, it's hard to tell what the behavior of the program is, which can lead to an incorrect reading of the code and to bugs.